lib/utils/net.go

```diff
@@ -17,11 +17,38 @@ limitations under the License.
 package utils
 
 import (
+	"context"
 	"net"
 
 	"github.com/gravitational/trace"
 )
 
+type webContextKey string
+
+const (
+	// ClientSrcAddrContextKey is context key for client source address
+	ClientSrcAddrContextKey webContextKey = "teleport-clientSrcAddrContextKey"
+	// ClientDstAddrContextKey is context key for client destination address
+	ClientDstAddrContextKey webContextKey = "teleport-clientDstAddrContextKey"
+)
+
+// ClientAddrContext is used by server to set incoming client's connection source and destination addresses to the context.
+func ClientAddrContext(ctx context.Context, src net.Addr, dst net.Addr) context.Context {
+	ctx = context.WithValue(ctx, ClientSrcAddrContextKey, src)
+	return context.WithValue(ctx, ClientDstAddrContextKey, dst)
+}
+
+// ClientAddrFromContext gets client source address and destination addresses from the context. If an address is
+// not present, nil will be returned
+func ClientAddrFromContext(ctx context.Context) (src net.Addr, dst net.Addr) {
```

Comment:
It's used to get client ip info on the server that is point of contact for client and then make this information available when it's needed. Like when client coming from WebUI, contacting our web server, we need to know this info to correctly create cert etc. In this PR it's also used to propagate client's IP when they are dialing app server. I'm also building upon it in upcoming pr (that enabled ip pinning enforcement for kube and DB) to propagate client ip from our kube's web server. It's not simple, but hopefully nobody actually need to think about these unless they are working specifically on ip pinning/ip propagation.

Reply:
Not necessarily for every new point of contact and I think not about those context functions. But in general if some new path was added and it requires client IP to be propagated, some work will need to be done in that direction. Mostly I think it will be about passing right `DialParams` or sending signed PROXY headers if it's direct connection. But end goal would be to have generic IP pinning enforcement check on everything incoming, like in auth middleware`withAuthenticatedUser` etc, so if IP propagation would be missing, it will just not work by default for a new pathway, so it will be easy to spot during development; basically having a good fail mode for ip pinning - if something's doesn't work right/missing it breaks and tells you and also it doesn't compromise security.